compiler/rustc_session/src/config/cfg.rs

```diff
@@ -83,6 +87,76 @@ impl<'a, T: Eq + Hash + Copy + 'a> Extend<&'a T> for ExpectedValues<T> {
     }
 }
 
+/// Disallow "builtin" cfgs from the CLI as they produce incoherent behavior
+pub(crate) fn disallow_cfgs(sess: &Session, user_cfgs: &Cfg) {
+    let disallow = |cfg: &(Symbol, Option<Symbol>), controlled_by| {
+        let cfg_name = cfg.0;
+        let cfg = if let Some(value) = cfg.1 {
+            format!(r#"{}="{}""#, cfg_name, value)
+        } else {
+            format!("{}", cfg_name)
+        };
+        sess.psess.opt_span_buffer_lint(
+            UNEXPECTED_BUILTIN_CFGS,
+            None,
+            ast::CRATE_NODE_ID,
+            BuiltinLintDiag::UnexpectedBuiltinCfg { cfg, cfg_name, controlled_by },
+        )
+    };
+
+    // We want to restrict setting cfgs that will produce "incoherent" behavior between
+    // the cfg and the "real" flag that sets it.
+    //
+    // The tests are in tests/ui/cfg/disallowed-cli-cfgs.rs.
+    //
+    // Not all "well known cfgs" can be disallowed, as they are set by external tools via the
+    // CLI, we therefore don't lint on them:
+    //
+    //  - test
+    //  - clippy
+    //  - doc
+    //  - doctest
+    //  - miri
+    //  - rustfmt
```

Comment:
Technically, this lint is about built-in cfgs and not "well known" cfgs, and all the cfgs in the list below are indeed built-in.
So we don't need to explain why we don't lint them (because they are not set by us).

`test` still needs to be explained though.